flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/TableEnvironment.java

```diff
@@ -0,0 +1,289 @@
+	 * @param configuredSink The configured {@link TableSink} to register.
+	 */
+	void registerTableSink(String name, TableSink<?> configuredSink);
+
+	/**
+	 * Scans a registered table and returns the resulting {@link Table}.
+	 *
+	 * <p>A table to scan must be registered in the TableEnvironment. It can be either directly
+	 * registered as DataStream, DataSet, or Table or as member of an {@link ExternalCatalog}.
+	 *
+	 * <p>Examples:
+	 *
+	 * <p>Scanning a directly registered table.
+	 * <pre>
+	 * {@code
+	 *   Table tab = tableEnv.scan("tableName");
+	 * }
+	 * </pre>
+	 *
+	 * <p>Scanning a table from a registered catalog.
+	 * <pre>
+	 * {@code
+	 *   Table tab = tableEnv.scan("catalogName", "dbName", "tableName");
+	 * }
+	 * </pre>
+	 *
+	 * @param tablePath The path of the table to scan.
+	 * @throws TableException if no table is found using the given table path.
+	 * @return The resulting {@link Table}.
+	 */
+	Table scan(String... tablePath) throws TableException;
+
+	/**
+	 * Creates a table source and/or table sink from a descriptor.
+	 *
+	 * <p>Descriptors allow for declaring the communication to external systems in an
+	 * implementation-agnostic way. The classpath is scanned for suitable table factories that match
+	 * the desired configuration.
+	 *
+	 * <p>The following example shows how to read from a connector using a JSON format and
+	 * registering a table source as "MyTable":
+	 *
+	 * <pre>
+	 * {@code
+	 *
+	 * tableEnv
+	 *   .connect(
+	 *     new ExternalSystemXYZ()
+	 *       .version("0.11"))
+	 *   .withFormat(
+	 *     new Json()
+	 *       .jsonSchema("{...}")
+	 *       .failOnMissingField(false))
+	 *   .withSchema(
+	 *     new Schema()
+	 *       .field("user-name", "VARCHAR").from("u_name")
+	 *       .field("count", "DECIMAL")
+	 *   .registerSource("MyTable");
+	 * }
+	 *</pre>
+	 *
+	 * @param connectorDescriptor connector descriptor describing the external system
+	 */
+	TableDescriptor connect(ConnectorDescriptor connectorDescriptor);
+
+	/**
+	 * Gets the names of all tables registered in this environment.
+	 *
+	 * @return A list of the names of all registered tables.
+	 */
+	String[] listTables();
+
+	/**
+	 * Gets the names of all functions registered in this environment.
+	 */
+	String[] listUserDefinedFunctions();
+
+	/**
+	 * Returns the AST of the specified Table API and SQL queries and the execution plan to compute
+	 * the result of the given {@link Table}.
+	 *
+	 * @param table The table for which the AST and execution plan will be returned.
+	 */
+	String explain(Table table);
+
+	/**
+	 * Returns completion hints for the given statement at the given cursor position.
+	 * The completion happens case insensitively.
+	 *
+	 * @param statement Partial or slightly incorrect SQL statement
+	 * @param position cursor position
+	 * @return completion hints that fit at the current cursor position
+	 */
+	String[] getCompletionHints(String statement, int position);
```

Comment:
I wonder if we should have this method in TableEnvironment interface... It is useful only for the sql-client and only for SQL (it makes no sense for Table API). SQL is (or will be ;) ) handled exclusively by the planner (maybe + access to Catalog, which will be a separate entity from `TableEnvironment` anyway). TableEnvironment just passes the string to the planner. I think we can base the sql client on the Planner interface directly in the long run.  

Therefore I would be up for dropping this method from `TableEnvironment` interface. WDYT @hequn8128 @twalthr ?

Reply:
Make sense. Maybe we can deprecate this method first and drop it in the next release, as it is public in the previous `TableEnvironment`. What do you think?

Reply:
I think this method is useful for anyone that uses the Table API as a library to build a service around it. I don't have a strong opinion here, but I think this could also stay as part of the interface.